root.go

```diff
@@ -204,14 +193,22 @@ func (kr *Root) updateChildEntry(c child) error {
 }
 
 func (kr *Root) Close() error {
-	nd, err := kr.GetDirectory().GetNode()
+	defer kr.cancel()
+
+	dir := kr.GetDirectory()
+	nd, err := dir.GetNode()
 	if err != nil {
 		return err
 	}
 
+	dir.lock.Lock()
+	defer dir.lock.Unlock()
+
 	if kr.repub != nil {
-		kr.repub.Update(nd.Cid())
-		return kr.repub.Close()
+		repub := kr.repub
+		kr.repub = nil
+		repub.Update(nd.Cid())
+		return repub.Close(kr.ctx)
 	}
```

Comment:
[I'm missing some context here and the change is definitely an improvement so feel free to ignore.] Why do we update in this case? I understand flushing in the `Close` of the file descriptor but I'm not sure I want to publish changes made directly to this structure (I'm not even sure who would manipulate the `Root` directly and even need a `Close` API except to stop the republisher).

Reply:
I am guessing is needed to ensure the root CID is published in the case where there were no updates to MFS.

As far as needing the Close() API... Part of graceful shutdown perhaps.  Did not intend to analyze that potential call path.  Just wanted to maintain functional equivalence here with safer code.